cmd/substreams-sink-postgres/inject_csv.go

```diff
@@ -0,0 +1,236 @@
+func (p sortedFilenames) Swap(i, j int)      { p[i], p[j] = p[j], p[i] }
+
+func s(str string) *string {
+	return &str
+}
+
+func extractFieldsFromFirstLine(ctx context.Context, filename string, store dstore.Store) ([]string, error) {
+	fl, err := store.OpenObject(ctx, filename)
+	if err != nil {
+		return nil, fmt.Errorf("opening csv: %w", err)
+	}
+	defer fl.Close()
+
+	r := csv.NewReader(fl)
+	out, err := r.Read()
+	if err != nil {
+		return nil, err
+	}
+
+	return out, nil
+}
+
+func (t *TableFiller) Run(ctx context.Context) error {
+	zlog.Info("table filler", zap.String("table", t.tblName))
+
+	loadFiles, err := injectFilesToLoad(t.in, t.tblName, t.stopBlockNum, t.startBlockNum)
+	if err != nil {
+		return fmt.Errorf("listing files: %w", err)
+	}
+
+	if len(loadFiles) == 0 {
+		return fmt.Errorf("no file to process")
+	}
+	dbFields, err := extractFieldsFromFirstLine(ctx, loadFiles[0], t.in)
+	if err != nil {
+		return fmt.Errorf("extracting fields from first csv line: %w", err)
+	}
+
+	zlog.Info("files to load",
+		zap.String("table", t.tblName),
+		zap.Int("file_count", len(loadFiles)),
+		//		zap.Int("pruned_file_count", len(prunedFilenames)),
+	)
+	prunedFilenames := loadFiles
+
+	for _, filename := range prunedFilenames {
+		zlog.Info("opening file", zap.String("file", filename))
+
+		if err := t.injectFile(ctx, filename, dbFields); err != nil {
+			return fmt.Errorf("failed to inject file %q: %w", filename, err)
+		}
+
+		// FIXME: set up a way to mark progress
+		//
+		//if err := t.markProgress(ctx, filename, time.Now()); err != nil {
+		//	return fmt.Errorf("failed to mark progress file %q: %w", filename, err)
+		//}
+	}
+
+	return nil
+}
+
+func (t *TableFiller) injectFile(ctx context.Context, filename string, dbFields []string) error {
+	fl, err := t.in.OpenObject(ctx, filename)
+	if err != nil {
+		return fmt.Errorf("opening csv: %w", err)
+	}
+	defer fl.Close()
+
+	query := fmt.Sprintf(`COPY %s.%s ("%s") FROM STDIN WITH (FORMAT CSV, HEADER)`, t.pqSchema, t.tblName, strings.Join(dbFields, `","`))
+	zlog.Info("loading file into sql", zap.String("filename", filename), zap.String("table_name", t.tblName), zap.Strings("db_fields", dbFields))
+
+	t0 := time.Now()
+
+	conn, err := t.pool.Acquire(ctx)
+	if err != nil {
+		return fmt.Errorf("pool acquire: %w", err)
+	}
+	defer conn.Release()
+
+	tag, err := conn.Conn().PgConn().CopyFrom(ctx, fl, query)
+	if err != nil {
+		return fmt.Errorf("failed COPY FROM for %q: %w", t.tblName, err)
+	}
+	count := tag.RowsAffected()
+	elapsed := time.Since(t0)
+	zlog.Info("loaded file into sql",
+		zap.String("filename", filename),
+		zap.String("table_name", t.tblName),
+		zap.Int64("rows_affected", count),
+		zap.Duration("elapsed", elapsed),
+	)
+
+	return nil
+}
+
+func injectFilesToLoad(inputStore dstore.Store, tableName string, stopBlockNum, desiredStartBlockNum uint64) (out []string, err error) {
+	err = inputStore.Walk(context.Background(), tableName+"/", func(filename string) (err error) {
```

Comment:
`context.Background()` should actually come in signature of `injectFilesToLoad`: `injectFilesToLoad(ctx context.Context, inputStore dstore.Store, ...`

Reply:
Done